### PR TITLE
Make _StorageBase.byteswap faster ( > 10000x)

### LINKED ISSUE #101690: _StrorageBase.byteswap() is too slow for large data
### 🐛 Describe the bug

The `_StorageBase.byteswap()` method is implemented in Python. Therefore, it is too slow for large data. I hit that it was not finished within 5 minutes when I loaded 560M parameters of a DNN model in a pickle file.

Possible solution is to write this method in C and to call it using pybind11.

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0a0+gite7681b5
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (s390x)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-71-generic-s390x-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:                    s390x
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Big Endian
CPU(s):                          8
On-line CPU(s) list:             0-7
Vendor ID:                       IBM/S390
Machine type:                    3931
Thread(s) per core:              1
Core(s) per socket:              1
Socket(s) per book:              1
Book(s) per drawer:              1
Drawer(s):                       8
CPU dynamic MHz:                 5200
CPU static MHz:                  5200
BogoMIPS:                        26315.00
Dispatching mode:                horizontal
Flags:                           esan3 zarch stfle msa ldisp eimm dfp edat etf3eh highgprs te vx vxd vxe gs vxe2 vxp sort dflt vxp2 nnpa sie
Hypervisor:                      z/VM 7.2.0
Hypervisor vendor:               IBM
Virtualization type:             full
L1d cache:                       1 MiB (8 instances)
L1i cache:                       1 MiB (8 instances)
L2 cache:                        256 MiB (8 instances)
L3 cache:                        256 MiB
NUMA node(s):                    1
NUMA node0 CPU(s):               0-7
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Not affected
Vulnerability Spectre v1:        Mitigation; __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; etokens
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.1.0a0+gite7681b5
[conda] Could not collect
```

### PR DESCRIPTION
**Pull request by kiszk**
This PR addresses #101690. This PR implement faster data elements swap in `_StorageBase` using C++ rather than using Python.

This PR helps such a situation that a large model saved on a little-endian machine will be loaded on a big-endian machine.

TODO:
- [x] Add test cases
- [x] Add performance comparison before and after the PR
- [ ] (Optional) Investigate further opportunities for performance improvements by [SIMDization](https://dev.to/wunk/fast-array-reversal-with-simd-j3p)

Fixes #101690

### CODE REVIEW

**Comment by vadimkantorov** [comment]
btw does torch.flip(uint8-tensor) implement something similar? (if inner dim is super small like 2/4/8 uint8)

**Comment by kiszk** [comment]
Thanks for your comment. IMHO, it looks similar, but a bit different in some cases.
```
python -c "import torch; print(torch.flip(torch.tensor([1,2,3,4,5,6,7,8], dtype=torch.uint8),[0]))"
tensor([8, 7, 6, 5, 4, 3, 2, 1], dtype=torch.uint8)
```

This PR would like to do for an 1-D array as follows:
```
elem_size=2: [1,2,3,4,5,6,7,8] -> [2,1,4,3,6,5,8,7]
elem_size=4: [1,2,3,4,5,6,7,8] -> [4,3,2,1,8,7,6,5]
elem_size=8: [1,2,3,4,5,6,7,8] -> [8,7,6,5,4,3,2,1]
```

Any comments?

**Comment by vadimkantorov** [comment]
Well, I mean that if you do `torch.tensor([0,1,2,3,4,5,6,7]).view(-1, 2).flip(-1).view(-1)`, it should do the thing. But it's not inplace and probably isn't optimized for these small 2/3/4/8 dimensions. I mean that it might make sense to improve torch.flip. This could then be used for inplace BGR<>RGB flips for (*, H, W, 3) images. Supporting inplace arbitrary shuffle would be even better :)

**Comment by kiszk** [comment]
Thanks, interesting. Since `flip()` does not support inplace version now, we would address it in another PR.

**Comment by kiszk** [comment]
Benchmark is that 64KB float32 elements (=256KB) are swapped. Measured 10 times and got an average.
Performance improvement is 176250x.

```
import timeit
import torch

buf = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15] * 64 * 256
storage = torch.storage.TypedStorage(buf, dtype=torch.uint8)._untyped_storage
print(timeit.timeit(lambda: storage.byteswap(torch.float32), number=10)/10, "sec.")
```

```
w/o this PR: 1.4130515402066521 sec.
w/  this PR: 8.233601693063975e-06 sec.
```

File: torch/storage.py [hunk 1]
```diff
@@ -29,7 +29,6 @@ class _StorageBase:
     def __init__(self, *args, **kwargs): ...  # noqa: E704
     def __len__(self) -> int: ...  # noqa: E704
     def __getitem__(self, idx): ...  # noqa: E704
-    def __setitem__(self, *args, **kwargs): ...  # noqa: E704
```
[the remark this answers is not in the document]
Comment (kiszk) [reply torch/storage.py:32]: It is intentionally. This code was introduced by https://github.com/pytorch/pytorch/pull/94503. This PR drops the code added by #94503.

So, since no one uses this `self[] = ...` now, I deleted this. Of course, we can still keep this. Which is preferable, keep or delete?

File: torch/csrc/StorageSharing.cpp [hunk 2]
```diff
@@ -622,6 +623,50 @@ PyObject* THPStorage_isShared(PyObject* self, PyObject* noargs) {
   }
 }
 
+PyObject* THPStorage_byteswap(PyObject* self, PyObject* args) {
```
[the remark this answers is not in the document]
Comment (kiszk) [reply torch/csrc/StorageSharing.cpp:626]: Good point. I considered that option, too. On the other hand, [a method table for `_StorageBase`](https://github.com/pytorch/pytorch/pull/101925/files#diff-5b86af2d42f697acbd39ea1d9ac7baffb854e09b661b36b2d6233fdf56fb305bR671) in `StorageMethods.cpp`. For ease of reading a file, I put the table and this function in the same file.

Is it better to split the places of the method table and this function?

File: torch/csrc/StorageSharing.cpp [hunk 3]
```diff
@@ -622,6 +623,50 @@ PyObject* THPStorage_isShared(PyObject* self, PyObject* noargs) {
...
+  THPUtils_assert(
+      elem_size == 1 || elem_size == 2 || elem_size == 4 || elem_size == 8,
+      "elem_size must be 1, 2, 4, or 8");
+
+  const auto& storage = THPStorage_Unpack(self);
+  const auto nbytes = static_cast<uint64_t>(storage.nbytes());
+
+  if (elem_size == 2) {
+    THPUtils_assert(
+        nbytes % 2 == 0, "the length of data is not a multiple of 2");
```
[the remark this answers is not in the document]
Comment (kiszk) [reply torch/csrc/StorageSharing.cpp]: Good catch, addressed

File: torch/csrc/StorageSharing.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/csrc/StorageSharing.cpp:626]: Hm, could we use this method table https://github.com/pytorch/pytorch/blob/415e84b6d5ab8b617111b4f603e48f2a180b72b2/torch/csrc/StorageMethods.cpp#L542

[the remark this answers is not in the document]
Comment (kiszk) [reply torch/csrc/StorageSharing.cpp:626]: Thank you for your suggestion. I will do this.

File: torch/csrc/StorageSharing.cpp [hunk 4]
```diff
@@ -12,6 +12,7 @@
 #include <torch/csrc/THP.h>
 #include <torch/csrc/autograd/utils/wrap_outputs.h>
 #include <torch/csrc/copy_utils.h>
+#include <torch/csrc/utils/byte_order.h>
```
Comment (mikaylagawarecki) [inline torch/csrc/StorageSharing.cpp:15]: small nit :)

**Review by mikaylagawarecki — APPROVED** [review]
LGTM, please fix the small nit!

**Comment by kiszk** [comment]
@mikaylagawarecki Thank you very much for your valuable comments.
